### PR TITLE
Extract content types from blob data

### PR DESCRIPTION
**Pull request by georgeclaghorn**
Active Storage relies on content types to expose features to categories of blobs. For example, variants are made available for blobs whose content types identify them as variable images.

Unfortunately, browsers provide unreliable content types for uploaded files. At Basecamp, we frequently observe browsers submitting image files with the `application/octet-stream` content type. If we relied on declared content types rather than extracting appropriate content types from file data, we would fail to properly display the misidentified image files. We’d fall back to a generic file icon instead of displaying the images themselves.

This PR brings content type detection to Active Storage via [Marcel](https://github.com/basecamp/marcel). Marcel uses [MimeMagic](https://github.com/minad/mimemagic/) to extract content types from file data. When MimeMagic chooses a vague content type for a file, Marcel attempts to pick a more specific one from its extension. 

Marcel is production-ready: we’ve used it in production for almost a year, and we’ve used magic number–based content type detection longer. It has an extensive test suite based in part on real edge cases we’ve identified in production.

### CODE REVIEW

File: activestorage/test/models/attachments_test.rb [hunk 1]
```diff
@@ -97,6 +97,29 @@ class ActiveStorage::AttachmentsTest < ActiveSupport::TestCase
...
+
+  test "identify newly-attached blob only once" do
+    blob = create_file_blob
+    assert blob.identified?
+
+    # The blob's backing file is a PNG image. Fudge its content type so we can tell if it's identified when we attach it.
+    blob.update! content_type: "application/octet-stream"
+
+    @user.avatar.attach blob
+    assert_equal "application/octet-stream", blob.content_type
```
Comment (kaspth) [inline activestorage/test/models/attachments_test.rb]: Shouldn’t this be `image/png`?

Comment (georgeclaghorn) [reply activestorage/test/models/attachments_test.rb]: No: we’re checking that the blob was *not* analyzed again.

File: activestorage/lib/active_storage/service/disk_service.rb [hunk 2]
```diff
@@ -9,10 +9,10 @@ module ActiveStorage
   # Wraps a local disk path as an Active Storage service. See ActiveStorage::Service for the generic API
   # documentation that applies to all services.
   class Service::DiskService < Service
-    attr_reader :root
+    attr_reader :root, :host
 
-    def initialize(root:)
-      @root = root
+    def initialize(root:, host:)
+      @root, @host = root, host
```
Comment (tenderlove) [inline activestorage/lib/active_storage/service/disk_service.rb]: Why does a disk service need a host?  It seems like a disk service wouldn't be over a network, so I don't understand this change.  Am I missing something?

Comment (georgeclaghorn) [reply activestorage/lib/active_storage/service/disk_service.rb]: It needs a host for the purpose of generating URLs.

Comment (tenderlove) [reply activestorage/lib/active_storage/service/disk_service.rb]: After reading the diff more closely, it looks like this is so that people can change the host for the generated URLs.  Maybe we should default the host when loading the config?  The reason I'm asking is because this change broke some existing configurations (ones without `host` specified).

Also we should probably fix the exception such that the filename for the malformed configuration is in the exception.

Comment (georgeclaghorn) [reply activestorage/lib/active_storage/service/disk_service.rb]: > it looks like this is so that people can change the host for the generated URLs

It’s because generated URLs previously didn't have any host at all. (`ActiveStorage::Service::DiskService#service_url` always returned a path.)  `ActiveStorage::Identification` (reasonably) expects `#service_url` to return a valid URL.

Comment (georgeclaghorn) [reply activestorage/lib/active_storage/service/disk_service.rb]: I apologize for not including this in the changelog, by the way. In retrospect, it deserved mention there.

Comment (tenderlove) [reply activestorage/lib/active_storage/service/disk_service.rb]: > I’m not sure what default we’d use. How would we know what host to use at boot time?

It looks like all of the generators generate `http://localhost:3000` as the default (which, technically isn't a host, but a URI).  Maybe that's a good start?  This code will give you the hostname of the current system:

```
$ ruby -rsocket -e'puts Socket.gethostname'
```

Since this is for storing / serving files locally, I'd expect localhost/3000 to be fine defaults.  Is there a reason why they aren't?

> I apologize for not including this in the changelog, by the way. In retrospect, it deserved mention there.

No problem!! 🤗

Comment (georgeclaghorn) [reply activestorage/lib/active_storage/service/disk_service.rb]: > It looks like all of the generators generate `http://localhost:3000` as the default (which, technically isn't a host, but a URI). Maybe that's a good start?

Sure. I’ll make that change in a bit.

Comment (georgeclaghorn) [reply activestorage/lib/active_storage/service/disk_service.rb]: Added a default in 6fb3ac1.
